sympy/functions/elementary/exponential.py

```diff
@@ -645,6 +645,23 @@ def fdiff(self, argindex=1):
         else:
             raise ArgumentIndexError(self, argindex)
 
+    def _eval_derivative(self, s):
+        if self.args[0].has(Pow):
+            a = Wild('a', exclude=[s])
+            b = Wild('b', exclude=[s, 0])
+            c = Wild('c', exclude=[s, 0])
+            d = Wild('d', exclude=[s])
+            e = Wild('e', exclude=[s])
+            f = Wild('f', exclude=[s, 0])
+            M = self.args[0].match(a+b*s+c*sqrt(d+e*s+f*s**2))
```

Comment:
This seems pretty expensive for a derivative calculation.

Reply:
It also seems a surprisingly specific.

Why is this change to differentiating logs needed?

Reply:
> Why is this change to differentiating logs needed?

When I changed the derivative of `acosh` in #23568, this test failed
https://github.com/sympy/sympy/blob/ff00553cf2a03ce048bf46c25c8a3cbe2a636c2c/sympy/integrals/tests/test_integrals.py#L1042-L1043
which means heurisch uses `acosh`'s derivative and compare it to the original expression.
Here I replaced `acosh` to `log`. For the same reason, the derivative of `log` will be compared to original expression. Without this change, heurisch can't deal with this test.

This change is tested in `test_log_diff`. I can't find a way to simplify the derivative of `log(b + 2*sqrt(c)*sqrt(a + b*x + c*x**2) + 2*c*x)` to `sqrt(c)/sqrt(a + b*x + c*x**2)`, and even if there exist a way, it's usually not recommended to post-simplify a step in SymPy code base.

> This seems pretty expensive for a derivative calculation.

1.10.1:
```py
In [3]: %timeit log(x**2+1).diff()
81.8 µs ± 459 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```
79dee2fd50166253be1eb8eb3577ab38f605218f:
```py
In [3]: %timeit log(x**2+1).diff()
38.7 ms ± 340 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
So, yes it's too expensive.
I'll push a change to not use that much `Wilds` and the time should be
```py
In [3]: %timeit log(x**2+1).diff()
821 µs ± 2.34 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Reply:
Unless there is a bug in the derivative of `log` I would rather just XFAIL the test than make any changes to the derivative of `log` regardless of speed.

Is the antiderivative in the test even correct though? With #23573 I get:
```python
In [32]: x = symbols('x')

In [33]: z = symbols('z', positive=True)

In [34]: F = -z**2*acosh(x/z)/2 + x*sqrt(x**2 - z**2)/2

In [35]: f = sqrt(x**2 - z**2)

In [36]: zero = F.diff(x) - f

In [37]: zero.subs({z:1,x:-2})
Out[37]: 
√3
──
3 
```

Reply:
The original antiderivative `-z**2*acosh(x/z)/2 + x*sqrt(x**2 - z**2)/2` was wrong. So I changed it to `x*sqrt(x**2 - z**2)/2 - z**2*log(x + sqrt(x**2 - z**2))/2` in this PR and the test can pass with the `log` derivative implementation.
Is XFAIL better even if it can be corrected? The zen of SymPy says
https://github.com/sympy/sympy/blob/645bbb06cd289165a9144d4c2c64bf17b664f8ac/sympy/this.py#L13-L14

Reply:
The reason I don't want to change log's derivative is because as far as I can tell it is correct. You haven't said that anything about it is incorrect so why is this change to log needed?

We can't just add random special cases everywhere. The correctness and simplicity of log is more important than evaluating any integral. If there is something that can be improved in heurisch or simplify or something then that can be considered but this is PR is trying to fix the problem in the wrong place.

Reply:
I think `simplify` can certainly be improved here although I'm not sure what's the best approach. Only `apart` seems capable of handling something like this with all of these square roots:
```python
In [67]: f = F.diff(x).subs(sqrt(a + b*x + c*x**2), z)

In [68]: f
Out[68]: 
     ⎛b      ⎞      
2⋅√c⋅⎜─ + c⋅x⎟      
     ⎝2      ⎠      
────────────── + 2⋅c
      z             
────────────────────
 b + 2⋅√c⋅z + 2⋅c⋅x 

In [69]: f.simplify()
Out[69]: 
√c⋅(b + 2⋅c⋅x) + 2⋅c⋅z
──────────────────────
z⋅(b + 2⋅√c⋅z + 2⋅c⋅x)

In [70]: ratsimp(f)
Out[70]: 
          3/2          
b⋅√c + 2⋅c   ⋅x + 2⋅c⋅z
───────────────────────
            2          
b⋅z + 2⋅√c⋅z  + 2⋅c⋅x⋅z

In [71]: radsimp(f)
Out[71]: 
 2           3/2        5/2  2      3/2  2
b ⋅√c + 4⋅b⋅c   ⋅x + 4⋅c   ⋅x  - 4⋅c   ⋅z 
──────────────────────────────────────────
     ⎛ 2                2  2        2⎞    
   z⋅⎝b  + 4⋅b⋅c⋅x + 4⋅c ⋅x  - 4⋅c⋅z ⎠    

In [72]: f.together()
Out[72]: 
√c⋅(b + 2⋅c⋅x) + 2⋅c⋅z
──────────────────────
z⋅(b + 2⋅√c⋅z + 2⋅c⋅x)

In [73]: f.expand()
Out[73]: 
                                     3/2                              
          b⋅√c                    2⋅c   ⋅x                 2⋅c        
─────────────────────── + ─────────────────────── + ──────────────────
            2                         2             b + 2⋅√c⋅z + 2⋅c⋅x
b⋅z + 2⋅√c⋅z  + 2⋅c⋅x⋅z   b⋅z + 2⋅√c⋅z  + 2⋅c⋅x⋅z                     

In [74]: f.expand(denom=True)
Out[74]: 
     ⎛b      ⎞      
2⋅√c⋅⎜─ + c⋅x⎟      
     ⎝2      ⎠      
────────────── + 2⋅c
      z             
────────────────────
 b + 2⋅√c⋅z + 2⋅c⋅x 

In [87]: factor(f)
Out[87]: 
          3/2          
b⋅√c + 2⋅c   ⋅x + 2⋅c⋅z
───────────────────────
 z⋅(b + 2⋅√c⋅z + 2⋅c⋅x)

In [75]: f.apart(z)
Out[75]: 
√c
──
z 
```
Note that if you substitute c=1 then almost all of the above will succeed:
```python
In [88]: f.subs(c, 1)
Out[88]: 
      ⎛b    ⎞
    2⋅⎜─ + x⎟
      ⎝2    ⎠
2 + ─────────
        z    
─────────────
b + 2⋅x + 2⋅z

In [89]: f.subs(c, 1).together()
Out[89]: 
1
─
z
```

Reply:
> why is this change to log needed?

[Wolfram alpha](https://www.wolframalpha.com/input?i=derivative+of+log%28b+%2B+2*sqrt%28c%29*sqrt%28a+%2B+b*x+%2B+c*x**2%29+%2B+2*c*x%29) already provides a simplified answer. I don't know the black magic behind it but it doesn't need users to manually call simplify.
Also, providing a simpler result does take more time but can also reduce processing time afterwards. Given `log(b + 2*sqrt(c)*sqrt(a + b*x + c*x**2) + 2*c*x).diff(x).simplify() == (sqrt(c)*(b + 2*c*x) + 2*c*sqrt(a + b*x + c*x**2))/(sqrt(a + b*x + c*x**2)*(b + 2*sqrt(c)*sqrt(a + b*x + c*x**2) + 2*c*x))`, it's not easy for user to simplify this result afterwards, and complexity accumulates fast.